design/rest-api-guideline.md

```diff
@@ -0,0 +1,389 @@
+
+#### Server Side Error Codes
+
+|Verb|Description|
+|---|---|
+|500|Indicates that the server could not fulfill the request due to some unexpected error|
+
+### Http Verbs
+
+Based on the REST architectural style, HTTP Verbs should be used as the means to communicate to APIs the operations to be performed
+on their resources. While HTTP verbs cover by themselves the most common operations you might want to look the [Controller Resource](#controllerresource) as an alternative on how to overcome any limitation imposed by the usage of HTTP verbs herein defined.
+
+The table below defines the HTTP verbs you should consider when exposing operations through your API and the circumstances where they apply.
+
+|Verb|Description|Idempotent|
+|---|---|---|
+|GET|Requests the current state of a resource|Yes|
+|POST|Creates a new resource or submit a command|No|
+|PUT|Replaces a resource state|Yes|
+|DELETE|Removes a resource|Yes|
+
+[TODO]: Shall we encourage PATCH for partial updates. That could make things easier for clients and avoid confusion when using PUT to update resources, where PUT
+should really be about replacing a resource, usually after a GET request to obtain the current resource representation.
+
+#### GET
+
+HTTP `GET` should be used to obtain current state of a resource. It is usually related with operations on resources that return either a list or a single representation, or start some action that does not change the resource state.
+
+##### Expected Status Codes
+
+It is expected that APIs responding to HTTP `GET` requests respect the following HTTP status codes to communicate the result of the operation:
+
+|Status Codes|
+|---|
+|200|
+|204|
+|307|
+|400|
+|401|
+|403|
+|404|
+|500|
+
+##### Notes on `201` Status Code and Actions using `GET`
+
+`GET` can represent some action supported by a resource (see [Controller Resource](#controllerresource) for more details) and eventually
+create resources as a consequence of the action. In this case, developers can use `201` on `GET` responses to communicate to clients the result of the operation.
+
+Under this circumstance, the developer should consider the same principles for `POST` requests especially in regards to managing
+conflicts.
+
+##### Notes on `204` Status Code and Actions using `GET`
+
+`GET` can represent some action supported by a resource (see [Controller Resource](#controllerresource) for more details) and eventually start
+some processing asynchronously without necessarily return a response body. In this case, developers can use `204` on `GET` responses to communicate to clients the result of the operation.
+
+#### POST
+
+HTTP `POST` should be used to create new resources. It is a non-safe and non-idempotent operation and clients should be aware of that.
+
+Successful `POST` requests should always result in a `201` status code and empty response. The response should also contain a reference
```

Comment:
Should the response be really empty? IMHO the representation of the newly created resource should be returned.

Reply:
The guideline states that a Location header should be returned pointing to the newly created resource. Considering that the location path includes the ID, clients can either use it to reach the resource that was created or extract the id from the location path in order to update the representation they already used to create the resource.

The point here is that we are avoiding unnecessary responses and saving network/processing.

Reply:
Maybe the question should be: "how often do you want to fetch a newly created resource?"
I'd say it's rather common that you create something (with some missing fields, like ID) and then you want to see the result with all the (generated) fields. And in this case we wouldn't be saving network - quite the opposite.

Reply:
In my previous reply, I mentioned how you could obtain the ID from the response. See the POST response in the guideline, isn't that enough for most cases? In our testsuite, for instance, there are places where this is enough.

When you create a resource you have the representation. If you want the ID is just a matter of checking the `Location` header and extract the ID. Or make another request if you really want to fetch the resource state.

Reply:
@vmuzikar I totally understand your point, the GitHub API follow a similar approach. But I would like us to try what Pedro suggested. This is recommend as best practice in https://restfulapi.net/http-status-201-created/.

Also, let's keep in mind that nothing is set in stone here. This draft will be reviewed by keycloak-dev mailing list and others. So if we reach the conclusion that's not gonna work, we can always revisit.

Reply:
@abstractj @pedroigor Ok, fair enough. :) Let's mark this as resolved.